File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Deque<int[]> q = new LinkedList<>();
        int w = 0;
        int time = 1;
        
        for (int truck : truck_weights) {
            while(!q.isEmpty() && time - q.peekFirst()[1] >= bridge_length || w + truck > weight) {
                int[] tt = q.pollFirst();
                w -= tt[0];
                time = tt[1] + bridge_length;
            }
            q.add(new int[]{truck, time});
            w += truck;
            time += 1;
        }
        return q.peekLast()[1] + bridge_length;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge, weight, trucks):
    q = deque()
    clock = 1
    q.append([trucks[0], clock])
    clock += 1
    current_weight = trucks[0]
    index = 1
    
    while index < len(trucks):
        if q and clock - q[0][1] > bridge:
            current_weight -= q.popleft()[0]
            
        if current_weight + trucks[index] <= weight:
            q.append([trucks[index], clock])
            current_weight += trucks[index]
            clock += 1
            index += 1
            
        else:
            t, c = q.popleft()
            current_weight -= t
            clock = c + bridge

            
    return q[-1][1] + bridge